DESIGN.md

```diff
@@ -19,19 +19,35 @@ This document contains the high level design of swiftly. Not all features have b
+
+In any case the delivery mechanism transitions to the initialization and these two parts form the complete installation process. Initialization is still needed since none of the delivery mechanisms typically modify the users' home directory and environment and it will be very useful to manage one workflow for everyone.
+
+The trigger for initialization is primarily swiftly's init subcommand, which can be invoked at any time to ensure that swiftly is initialized for the user, guiding them through the process if it isn't. This also serves as an entry point to upgrade if the user has an older swiftly installation. The init subcommand is responsible for setting up an initial directory structure for storing toolchains, the swiftly configuration `config.json`, and a shell environment `env.sh` (or similar for the user's shell). Operating system depenencies are checked, and if any are missing then the user is prompted to either install them automatically, or they are given the package installation commands to do so themselves. Note that the init can be invoked without any prompts, with supporting flags to provide needed information, to support scripted installations.
+
+```
+swiftly init
+```
+
+Since an end user may not be aware that swiftly has an initialiation step, its possible that they will invoke it in an uninitialized state on their account. If this happens then swiftly validates its own installation every time it is invoked and helps the user to discover the init subcommand. A system package, or website, can use an accelerator to help guide the user to the init subcommand by naming the swiftly binary as `swiftly-init*`. Swiftly discovers that it is being invoked in this way and automatically runs the initialization right away. If the system package installs swiftly-init in a location that's on the user's default path, tab completion of `swiftly` will usually complete to the full name and the user is directed to run the initialization by default. However, if swiftly is already on the user's path then the completion will be unintrusive.
+
+```
+swiftly-init-0.4.0 # Runs 'swiftly init' subcommand to initialize swiftly
+```
+
+### Swiftly upgrades
+
+It is possible that initialization will detect an existing swiftly installation at an older version, which then performs an upgrade procedure. If the existing installation is newer (using the `--version flag`) then the initialization is aborted. This gives the user the opportunity to upgrade to a newer version of swiftly by manually downloading it from a trusted website or ad-hoc test a pre-release.
```

Comment:
I would imagine that the case we should be solving for here is not that there are multiple binaries on the system, but rather that a newer swiftly binary was dropped in when an older configuration existed. In that case, we'd detect this via some marker in the `config.json`, rather than by invoking `--version`. I don't think we'll want swiftly combing the PATH for other versions of swiftly really, and we can just rely on the contents of `config.json`.